onnx/defs/nn/defs.cc

```diff
@@ -2487,4 +2487,201 @@ ONNX_OPERATOR_SET_SCHEMA(
         )ONNX"
         ));
 
+static const char* LayerNormalization_ver16_doc = R"DOC(
+      This is layer normalization defined in ONNX as function.
+      The overall computation can be split into two stages.
+      The first stage is standardization, which makes the
+      normalized elements have zero mean and unit variances.
+      The computation required by standardization can be
+      described by the following equations.
+      ```
+      Mean = ReduceMean<axes=normalized_axes>(X)
+      D = Sub(X, Mean)
+      DD = Mul(Diff, Diff)
+      Var = ReduceMean<axes=axes>(DD)
+      VarEps = Add(Var, epsilon)
+      StdDev = Sqrt(VarEps)
+      InvStdDev = Reciprocal(StdDev)
+      Normalized = Mul(D, InvStdDev)
+      ```
+      where `normalized_axes` is `[axis, ..., rank of X - 1]`.
+      The variables `Var` and `StdDev` stand for variance and
+      standard deviation, respectively. The second output is
+      `Mean` and the last one is `InvStdDev`.
+      Depending on `stash_type` attribute, the actual computation may
+      happen in different floating-point precision.
+      For example, if `stash_type` is 1, this operator may cast
+      all input variables to 32-bit float, perform the computation, and
+      finally cast `Normalized` back to the original type of `X`.
+      The second stage then scales and shifts the outcome of
+      stage one using
+      ```
+      NormalizedScaled = Mul(Normalized, Scale)
+      Y = Add(NormalizedScaled, B)
+      ```
+      All equations are in [this syntax](https://github.com/onnx/onnx/blob/main/docs/Syntax.md).
+      The same variable (i.e., input, output, and attribute) uses
+      the same name in the equations above and this operator's definition.
+      Let `d[i]` indicate the i-th dimension of `X`.
+      If `X`'s shape is `[d[0], ..., d[axis-1], d[axis], ..., d[rank-1]]`,
+      the shape of `Mean` and `InvStdDev` is `[d[0], ..., d[axis-1], 1, ..., 1]`.
+      `Y` and `X` have the same shape.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    LayerNormalization,
+    16,
+    OpSchema()
+        .SetDoc(LayerNormalization_ver16_doc)
+        .Attr("axis",
+              "The first normalization dimension. If rank(X) is r, axis' allowed range is [-r, r]. "
+              "Negative value means counting dimensions from the back.",
+              AttributeProto::INT,
+              static_cast<int64_t>(-1))
+        .Attr("epsilon",
+              "The epsilon value to use to avoid division by zero.",
+              AttributeProto::FLOAT,
+              1e-5f)
+        .Attr("stash_type",
+              "Type of Mean and InvStdDev. This also specifies stage one's computation precision.",
+              AttributeProto::INT,
+              static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_FLOAT))
+        .AllowUncheckedAttributes()
+        .Input(0, "X", "Tensor to be normalized.", "T")
+        .Input(1, "Scale", "Scale tensor.", "T")
```

Comment:
The definition of BatchNorm went through multiple revisions because there was a need to allow different precisions for different inputs. Is it fine if X/Scale/B are all constrained to be same type?

Reply:
Layer normalization is very similar to batch normalization (major difference is the normalized axes). As batch normalization have shown the need of using difference precision for `mean` and `variance` tensors, I prefer to keep this as is for layer normalization's `mean` and `variance`. In addition, mix-precision training for language model is popular enough to support extra types. For inference, I guess mix-precision is also helpful to compress the trained models. Does it sound ok?

Reply:
Discussed with Wei-Sheng. Summary: the inputs will be of same type for now, as there is no use-case yet of using mixed-precision for inputs.